### PR TITLE
access logging: support custom tag on ALS

### LINKED ISSUE #18910: access logging custom tagging feature
In tracing implementation, envoy has [custom tag](https://www.envoyproxy.io/docs/envoy/latest/api-v3/type/tracing/v3/custom_tag.proto) feature. I think that we need a similar concept in access logging, which helps to annotate logs and provides an easier way to arrange collected logs.

### PR DESCRIPTION
**Pull request by Shikugawa**
Signed-off-by: Shikugawa <rei@tetrate.io>

Commit Message: access logging: support custom tag on ALS
Additional Description: Close #18910. This PR adds custom tag feature that is similar to tracing span's. In this PR, we support this for ALS only.
Risk Level: Low
Testing: Unit
Docs Changes: Required
Release Notes:
Platform Specific Features:
[Optional Runtime guard:]
[Optional Fixes #Issue]
[Optional Fixes commit #PR or SHA]
[Optional Deprecated:]
[Optional [API Considerations](https://github.com/envoyproxy/envoy/blob/main/api/review_checklist.md):]

### CODE REVIEW

**Comment by lizan** [comment]
@mathetake first pass?

File: api/envoy/data/accesslog/v3/accesslog.proto [hunk 1]
```diff
@@ -183,6 +183,9 @@ message AccessLogCommon {
   // state serialized to any message other than `google.protobuf.Any` it will be packed into
   // `google.protobuf.Any`.
   map<string, google.protobuf.Any> filter_state_objects = 21;
+
+  // A list of custom tags with unique tag name to create tags for the logs.
```
Comment (mathetake) [inline api/envoy/data/accesslog/v3/accesslog.proto:187]: Could you elaborate on how "tag" means here? new comer won't understand what this actually does. Maybe better add explanation annotation, etc?

File: envoy/tracing/trace_driver.h [hunk 2]
```diff
@@ -5,6 +5,7 @@
 #include <vector>
 
 #include "envoy/common/pure.h"
+#include "envoy/data/accesslog/v3/accesslog.pb.h"
```
Comment (mathetake) [inline envoy/tracing/trace_driver.h:8]: so this introduces the dependency on access_logs *in* tracing/trace_driver? I understand how this ends up here but seems a bit weird to me.

File: envoy/tracing/trace_driver.h [hunk 3]
```diff
@@ -45,7 +46,15 @@ class CustomTag {
    * @param span the active span.
    * @param ctx the custom tag context.
    */
-  virtual void apply(Span& span, const CustomTagContext& ctx) const PURE;
+  virtual void applySpan(Span& span, const CustomTagContext& ctx) const PURE;
+
+  /**
+   * Get string tag value from various type of custom tags.
```
Comment (mathetake) [inline envoy/tracing/trace_driver.h:52]: what do you mean here by "various type"?

**Comment by mathetake** [comment]
Made drive-by comments :D

File: envoy/tracing/trace_driver.h (hunk 2 above)
Comment (Shikugawa) [reply envoy/tracing/trace_driver.h:8]: I also considered this dependency seems weird, then submitted refactor PR. https://github.com/envoyproxy/envoy/pull/18942#issuecomment-965378956
But, maintainers referred to me utilizing the current API.

**Comment by Shikugawa** [comment]
@lizan Could you take a look?

**Comment by ggreenway** [comment]
@lizan ping

**Comment by wu-sheng** [comment]
@lizan Could you help on this?

File: source/extensions/access_loggers/grpc/BUILD [hunk 4]
```diff
@@ -51,6 +51,7 @@ envoy_cc_library(
         "//source/common/network:utility_lib",
         "//source/common/stream_info:stream_info_lib",
         "//source/common/stream_info:utility_lib",
+        "//source/common/tracing:http_tracer_lib",
```
Comment (lizan) [inline source/extensions/access_loggers/grpc/BUILD:54]: I don't think gRPC access logger should pull whole HTTP tracer as a dependency. Can you pull `CustomTagBase` and related class to different file and an `envoy_cc_library`?

Comment (Shikugawa) [reply source/extensions/access_loggers/grpc/BUILD]: done

**Comment by Shikugawa** [comment]
@envoyproxy/envoy-maintainers Could anyone help this?

**Review by lizan — COMMENTED** [review]
Can you extract the custom tag refactor to another PR? otherwise looks good to me.

**Review by lizan — COMMENTED** [review]
/retest

**Review by lizan — APPROVED** [review]

**Comment by Shikugawa** [comment]
@snowp friendly ping

**Review by snowp — APPROVED** [review]
Thanks!

**Comment by wu-sheng** [comment]
Great to see this gets merged.

**Pull request merged**